File: app/src/main/java/tw/firemaples/onscreenocr/utils/SignatureUtil.java
package tw.firemaples.onscreenocr.utils;

import android.content.Context;
import android.content.pm.PackageManager;
import android.content.pm.Signature;
import android.util.Base64;

import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;

/**
 * Created by firemaples on 07/07/2017.
 */

public class SignatureUtil {
    private static final String SIGNATURE_SHA = "gK7G8ZtYKeXYMsWzOZ0wP+rTP50=\n";
    private static final int SIGNATURE_HASH = 116585913;

    public static boolean validateSignature(Context context) throws PackageManager.NameNotFoundException, NoSuchAlgorithmException {
        String currentSignatureSHA = getCurrentSignatureSHA(context);
        if (currentSignatureSHA == null) {
            return false;
        } else {
            return currentSignatureSHA.replaceAll("\n", "").equals(SIGNATURE_SHA.replaceAll("\n", ""));
        }
    }

    public static String getCurrentSignatureSHA(Context context) throws NoSuchAlgorithmException, PackageManager.NameNotFoundException {
        Signature signature = getCurrentSignature(context);
        if (signature == null) {
            return null;
        }
        byte[] signatureBytes = signature.toByteArray();
        MessageDigest md = MessageDigest.getInstance("SHA");
        md.update(signatureBytes);
        return Base64.encodeToString(md.digest(), Base64.DEFAULT);
    }

    public static int getCurrentSignatureHashCode(Context context) throws PackageManager.NameNotFoundException {
        Signature signature = getCurrentSignature(context);
        if (signature == null) {
            return -1;
        }
        return signature.hashCode();
    }

    public static Signature getCurrentSignature(Context context) throws PackageManager.NameNotFoundException {
        Signature[] signatures = context.getPackageManager().getPackageInfo(context.getPackageName(), PackageManager.GET_SIGNATURES).signatures;
        if (signatures != null && signatures.length > 0) {
            return signatures[0];
        }
        return null;
    }
}

File: main/src/main/java/tw/firemaples/onscreenocr/utils/SignatureUtil.kt
package tw.firemaples.onscreenocr.utils

import android.content.Context
import android.content.pm.PackageManager
import android.content.pm.Signature
import android.util.Base64
import java.security.MessageDigest
import java.security.NoSuchAlgorithmException

class SignatureUtil {
    companion object {
        private const val SIGNATURE_SHA = "gK7G8ZtYKeXYMsWzOZ0wP+rTP50=\n"
        private const val SIGNATURE_HASH = 116585913

        @Throws(PackageManager.NameNotFoundException::class, NoSuchAlgorithmException::class)
        fun validateSignature(context: Context): Boolean {
            val currentSignatureSHA = getCurrentSignatureSHA(context) ?: return false

            return currentSignatureSHA.replace("\n", "") ==
                    SIGNATURE_SHA.replace("\n", "")
        }

        @Throws(NoSuchAlgorithmException::class, PackageManager.NameNotFoundException::class)
        fun getCurrentSignatureSHA(context: Context): String? {
            val signature = getCurrentSignature(context) ?: return null
            val md = MessageDigest.getInstance("SHA").apply {
                update(signature.toByteArray())
            }

            return Base64.encodeToString(md.digest(), Base64.DEFAULT)
        }

        @Throws(PackageManager.NameNotFoundException::class)
        fun getCurrentSignatureHashCode(context: Context): Int =
            getCurrentSignature(context)?.hashCode() ?: -1

        @Throws(PackageManager.NameNotFoundException::class)
        fun getCurrentSignature(context: Context): Signature? {
            val signatures = context.packageManager
                .getPackageInfo(context.packageName, PackageManager.GET_SIGNATURES).signatures
            
            return if (!signatures.isNullOrEmpty()) signatures[0] else null
        }
    }
}
